// @mcp-use/modelcontextprotocol-sdk/dist/cjs/shared/auth.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OAuthTokenRevocationRequestSchema = exports.OAuthClientRegistrationErrorSchema = exports.OAuthClientInformationFullSchema = exports.OAuthClientInformationSchema = exports.OAuthClientMetadataSchema = exports.OptionalSafeUrlSchema = exports.OAuthErrorResponseSchema = exports.OAuthTokensSchema = exports.OpenIdProviderDiscoveryMetadataSchema = exports.OpenIdProviderMetadataSchema = exports.OAuthMetadataSchema = exports.OAuthProtectedResourceMetadataSchema = exports.SafeUrlSchema = void 0;
const v4_1 = __importDefault(require("zod/v4"));
/**
 * Reusable URL validation that disallows javascript: scheme
 */
exports.SafeUrlSchema = v4_1.default
    .url()
    .superRefine((val, ctx) => {
    if (!URL.canParse(val)) {
        ctx.addIssue({
            code: v4_1.default.ZodIssueCode.custom,
            message: 'URL must be parseable',
            fatal: true
        });
        return v4_1.default.NEVER;
    }
})
    .refine(url => {
    const u = new URL(url);
    return u.protocol !== 'javascript:' && u.protocol !== 'data:' && u.protocol !== 'vbscript:';
}, { message: 'URL cannot use javascript:, data:, or vbscript: scheme' });
/**
 * RFC 9728 OAuth Protected Resource Metadata
 */
exports.OAuthProtectedResourceMetadataSchema = v4_1.default.looseObject({
    resource: v4_1.default.string().url(),
    authorization_servers: v4_1.default.array(exports.SafeUrlSchema).optional(),
    jwks_uri: v4_1.default.string().url().optional(),
    scopes_supported: v4_1.default.array(v4_1.default.string()).optional(),
    bearer_methods_supported: v4_1.default.array(v4_1.default.string()).optional(),
    resource_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    resource_name: v4_1.default.string().optional(),
    resource_documentation: v4_1.default.string().optional(),
    resource_policy_uri: v4_1.default.string().url().optional(),
    resource_tos_uri: v4_1.default.string().url().optional(),
    tls_client_certificate_bound_access_tokens: v4_1.default.boolean().optional(),
    authorization_details_types_supported: v4_1.default.array(v4_1.default.string()).optional(),
    dpop_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    dpop_bound_access_tokens_required: v4_1.default.boolean().optional()
});
/**
 * RFC 8414 OAuth 2.0 Authorization Server Metadata
 */
exports.OAuthMetadataSchema = v4_1.default.looseObject({
    issuer: v4_1.default.string(),
    authorization_endpoint: exports.SafeUrlSchema,
    token_endpoint: exports.SafeUrlSchema,
    registration_endpoint: exports.SafeUrlSchema.optional(),
    scopes_supported: v4_1.default.array(v4_1.default.string()).optional(),
    response_types_supported: v4_1.default.array(v4_1.default.string()),
    response_modes_supported: v4_1.default.array(v4_1.default.string()).optional(),
    grant_types_supported: v4_1.default.array(v4_1.default.string()).optional(),
    token_endpoint_auth_methods_supported: v4_1.default.array(v4_1.default.string()).optional(),
    token_endpoint_auth_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    service_documentation: exports.SafeUrlSchema.optional(),
    revocation_endpoint: exports.SafeUrlSchema.optional(),
    revocation_endpoint_auth_methods_supported: v4_1.default.array(v4_1.default.string()).optional(),
    revocation_endpoint_auth_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    introspection_endpoint: v4_1.default.string().optional(),
    introspection_endpoint_auth_methods_supported: v4_1.default.array(v4_1.default.string()).optional(),
    introspection_endpoint_auth_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    code_challenge_methods_supported: v4_1.default.array(v4_1.default.string()).optional(),
    client_id_metadata_document_supported: v4_1.default.boolean().optional()
});
/**
 * OpenID Connect Discovery 1.0 Provider Metadata
 * see: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
 */
exports.OpenIdProviderMetadataSchema = v4_1.default.looseObject({
    issuer: v4_1.default.string(),
    authorization_endpoint: exports.SafeUrlSchema,
    token_endpoint: exports.SafeUrlSchema,
    userinfo_endpoint: exports.SafeUrlSchema.optional(),
    jwks_uri: exports.SafeUrlSchema,
    registration_endpoint: exports.SafeUrlSchema.optional(),
    scopes_supported: v4_1.default.array(v4_1.default.string()).optional(),
    response_types_supported: v4_1.default.array(v4_1.default.string()),
    response_modes_supported: v4_1.default.array(v4_1.default.string()).optional(),
    grant_types_supported: v4_1.default.array(v4_1.default.string()).optional(),
    acr_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    subject_types_supported: v4_1.default.array(v4_1.default.string()),
    id_token_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()),
    id_token_encryption_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    id_token_encryption_enc_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    userinfo_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    userinfo_encryption_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    userinfo_encryption_enc_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    request_object_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    request_object_encryption_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    request_object_encryption_enc_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    token_endpoint_auth_methods_supported: v4_1.default.array(v4_1.default.string()).optional(),
    token_endpoint_auth_signing_alg_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    display_values_supported: v4_1.default.array(v4_1.default.string()).optional(),
    claim_types_supported: v4_1.default.array(v4_1.default.string()).optional(),
    claims_supported: v4_1.default.array(v4_1.default.string()).optional(),
    service_documentation: v4_1.default.string().optional(),
    claims_locales_supported: v4_1.default.array(v4_1.default.string()).optional(),
    ui_locales_supported: v4_1.default.array(v4_1.default.string()).optional(),
    claims_parameter_supported: v4_1.default.boolean().optional(),
    request_parameter_supported: v4_1.default.boolean().optional(),
    request_uri_parameter_supported: v4_1.default.boolean().optional(),
    require_request_uri_registration: v4_1.default.boolean().optional(),
    op_policy_uri: exports.SafeUrlSchema.optional(),
    op_tos_uri: exports.SafeUrlSchema.optional(),
    client_id_metadata_document_supported: v4_1.default.boolean().optional()
});
/**
 * OpenID Connect Discovery metadata that may include OAuth 2.0 fields
 * This schema represents the real-world scenario where OIDC providers
 * return a mix of OpenID Connect and OAuth 2.0 metadata fields
 */
exports.OpenIdProviderDiscoveryMetadataSchema = v4_1.default.object({
    ...exports.OpenIdProviderMetadataSchema.shape,
    ...exports.OAuthMetadataSchema.pick({
        code_challenge_methods_supported: true
    }).shape
});
/**
 * OAuth 2.1 token response
 */
exports.OAuthTokensSchema = v4_1.default
    .object({
    access_token: v4_1.default.string(),
    id_token: v4_1.default.string().optional(), // Optional for OAuth 2.1, but necessary in OpenID Connect
    token_type: v4_1.default.string(),
    expires_in: v4_1.default.coerce.number().optional(),
    scope: v4_1.default.string().optional(),
    refresh_token: v4_1.default.string().optional()
})
    .strip();
/**
 * OAuth 2.1 error response
 */
exports.OAuthErrorResponseSchema = v4_1.default.object({
    error: v4_1.default.string(),
    error_description: v4_1.default.string().optional(),
    error_uri: v4_1.default.string().optional()
});
/**
 * Optional version of SafeUrlSchema that allows empty string for retrocompatibility on tos_uri and logo_uri
 */
exports.OptionalSafeUrlSchema = exports.SafeUrlSchema.optional().or(v4_1.default.literal('').transform(() => undefined));
/**
 * RFC 7591 OAuth 2.0 Dynamic Client Registration metadata
 */
exports.OAuthClientMetadataSchema = v4_1.default
    .object({
    redirect_uris: v4_1.default.array(exports.SafeUrlSchema),
    token_endpoint_auth_method: v4_1.default.string().optional(),
    grant_types: v4_1.default.array(v4_1.default.string()).optional(),
    response_types: v4_1.default.array(v4_1.default.string()).optional(),
    client_name: v4_1.default.string().optional(),
    client_uri: exports.SafeUrlSchema.optional(),
    logo_uri: exports.OptionalSafeUrlSchema,
    scope: v4_1.default.string().optional(),
    contacts: v4_1.default.array(v4_1.default.string()).optional(),
    tos_uri: exports.OptionalSafeUrlSchema,
    policy_uri: v4_1.default.string().optional(),
    jwks_uri: exports.SafeUrlSchema.optional(),
    jwks: v4_1.default.any().optional(),
    software_id: v4_1.default.string().optional(),
    software_version: v4_1.default.string().optional(),
    software_statement: v4_1.default.string().optional()
})
    .strip();
/**
 * RFC 7591 OAuth 2.0 Dynamic Client Registration client information
 */
exports.OAuthClientInformationSchema = v4_1.default
    .object({
    client_id: v4_1.default.string(),
    client_secret: v4_1.default.string().optional(),
    client_id_issued_at: v4_1.default.number().optional(),
    client_secret_expires_at: v4_1.default.number().optional()
})
    .strip();
/**
 * RFC 7591 OAuth 2.0 Dynamic Client Registration full response (client information plus metadata)
 */
exports.OAuthClientInformationFullSchema = exports.OAuthClientMetadataSchema.merge(exports.OAuthClientInformationSchema);
/**
 * RFC 7591 OAuth 2.0 Dynamic Client Registration error response
 */
exports.OAuthClientRegistrationErrorSchema = v4_1.default
    .object({
    error: v4_1.default.string(),
    error_description: v4_1.default.string().optional()
})
    .strip();
/**
 * RFC 7009 OAuth 2.0 Token Revocation request
 */
exports.OAuthTokenRevocationRequestSchema = v4_1.default
    .object({
    token: v4_1.default.string(),
    token_type_hint: v4_1.default.string().optional()
})
    .strip();
