// @mcp-use/modelcontextprotocol-sdk/dist/cjs/examples/server/mcpServerOutputSchema.js
#!/usr/bin/env node
"use strict";
/**
 * Example MCP server using the high-level McpServer API with outputSchema
 * This demonstrates how to easily create tools with structured output
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const v4_1 = __importDefault(require("zod/v4"));
const mcp_js_1 = require("../../server/mcp.js");
const stdio_js_1 = require("../../server/stdio.js");
const server = new mcp_js_1.McpServer({
    name: 'mcp-output-schema-high-level-example',
    version: '1.0.0'
});
// Define a tool with structured output - Weather data
server.registerTool('get_weather', {
    description: 'Get weather information for a city',
    inputSchema: {
        city: v4_1.default.string().describe('City name'),
        country: v4_1.default.string().describe('Country code (e.g., US, UK)')
    },
    outputSchema: {
        temperature: v4_1.default.object({
            celsius: v4_1.default.number(),
            fahrenheit: v4_1.default.number()
        }),
        conditions: v4_1.default.enum(['sunny', 'cloudy', 'rainy', 'stormy', 'snowy']),
        humidity: v4_1.default.number().min(0).max(100),
        wind: v4_1.default.object({
            speed_kmh: v4_1.default.number(),
            direction: v4_1.default.string()
        })
    }
}, async ({ city, country }) => {
    // Parameters are available but not used in this example
    void city;
    void country;
    // Simulate weather API call
    const temp_c = Math.round((Math.random() * 35 - 5) * 10) / 10;
    const conditions = ['sunny', 'cloudy', 'rainy', 'stormy', 'snowy'][Math.floor(Math.random() * 5)];
    const structuredContent = {
        temperature: {
            celsius: temp_c,
            fahrenheit: Math.round(((temp_c * 9) / 5 + 32) * 10) / 10
        },
        conditions,
        humidity: Math.round(Math.random() * 100),
        wind: {
            speed_kmh: Math.round(Math.random() * 50),
            direction: ['N', 'NE', 'E', 'SE', 'S', 'SW', 'W', 'NW'][Math.floor(Math.random() * 8)]
        }
    };
    return {
        content: [
            {
                type: 'text',
                text: JSON.stringify(structuredContent, null, 2)
            }
        ],
        structuredContent
    };
});
async function main() {
    const transport = new stdio_js_1.StdioServerTransport();
    await server.connect(transport);
    console.error('High-level Output Schema Example Server running on stdio');
}
main().catch(error => {
    console.error('Server error:', error);
    process.exit(1);
});
