// @mcp-use/modelcontextprotocol-sdk/dist/cjs/types.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskStatusNotificationParamsSchema = exports.CreateTaskResultSchema = exports.TaskSchema = exports.PaginatedResultSchema = exports.PaginatedRequestSchema = exports.PaginatedRequestParamsSchema = exports.ProgressNotificationSchema = exports.ProgressNotificationParamsSchema = exports.ProgressSchema = exports.PingRequestSchema = exports.isInitializedNotification = exports.InitializedNotificationSchema = exports.InitializeResultSchema = exports.ServerCapabilitiesSchema = exports.isInitializeRequest = exports.InitializeRequestSchema = exports.InitializeRequestParamsSchema = exports.ClientCapabilitiesSchema = exports.ServerTasksCapabilitySchema = exports.ClientTasksCapabilitySchema = exports.ImplementationSchema = exports.BaseMetadataSchema = exports.IconsSchema = exports.IconSchema = exports.CancelledNotificationSchema = exports.CancelledNotificationParamsSchema = exports.EmptyResultSchema = exports.JSONRPCMessageSchema = exports.isJSONRPCError = exports.JSONRPCErrorSchema = exports.ErrorCode = exports.isJSONRPCResponse = exports.JSONRPCResponseSchema = exports.isJSONRPCNotification = exports.JSONRPCNotificationSchema = exports.isJSONRPCRequest = exports.JSONRPCRequestSchema = exports.RequestIdSchema = exports.ResultSchema = exports.NotificationSchema = exports.RequestSchema = exports.RelatedTaskMetadataSchema = exports.TaskCreationParamsSchema = exports.CursorSchema = exports.ProgressTokenSchema = exports.JSONRPC_VERSION = exports.RELATED_TASK_META_KEY = exports.SUPPORTED_PROTOCOL_VERSIONS = exports.DEFAULT_NEGOTIATED_PROTOCOL_VERSION = exports.LATEST_PROTOCOL_VERSION = void 0;
exports.ListToolsRequestSchema = exports.ToolSchema = exports.ToolExecutionSchema = exports.ToolAnnotationsSchema = exports.PromptListChangedNotificationSchema = exports.GetPromptResultSchema = exports.PromptMessageSchema = exports.ContentBlockSchema = exports.ResourceLinkSchema = exports.EmbeddedResourceSchema = exports.ToolUseContentSchema = exports.AudioContentSchema = exports.ImageContentSchema = exports.TextContentSchema = exports.GetPromptRequestSchema = exports.GetPromptRequestParamsSchema = exports.ListPromptsResultSchema = exports.ListPromptsRequestSchema = exports.PromptSchema = exports.PromptArgumentSchema = exports.ResourceUpdatedNotificationSchema = exports.ResourceUpdatedNotificationParamsSchema = exports.UnsubscribeRequestSchema = exports.UnsubscribeRequestParamsSchema = exports.SubscribeRequestSchema = exports.SubscribeRequestParamsSchema = exports.ResourceListChangedNotificationSchema = exports.ReadResourceResultSchema = exports.ReadResourceRequestSchema = exports.ReadResourceRequestParamsSchema = exports.ResourceRequestParamsSchema = exports.ListResourceTemplatesResultSchema = exports.ListResourceTemplatesRequestSchema = exports.ListResourcesResultSchema = exports.ListResourcesRequestSchema = exports.ResourceTemplateSchema = exports.ResourceSchema = exports.AnnotationsSchema = exports.RoleSchema = exports.BlobResourceContentsSchema = exports.TextResourceContentsSchema = exports.ResourceContentsSchema = exports.CancelTaskResultSchema = exports.CancelTaskRequestSchema = exports.ListTasksResultSchema = exports.ListTasksRequestSchema = exports.GetTaskPayloadRequestSchema = exports.GetTaskResultSchema = exports.GetTaskRequestSchema = exports.TaskStatusNotificationSchema = void 0;
exports.ListRootsRequestSchema = exports.RootSchema = exports.CompleteResultSchema = exports.CompleteRequestSchema = exports.CompleteRequestParamsSchema = exports.PromptReferenceSchema = exports.ResourceReferenceSchema = exports.ResourceTemplateReferenceSchema = exports.ElicitResultSchema = exports.ElicitationCompleteNotificationSchema = exports.ElicitationCompleteNotificationParamsSchema = exports.ElicitRequestSchema = exports.ElicitRequestParamsSchema = exports.ElicitRequestURLParamsSchema = exports.ElicitRequestFormParamsSchema = exports.PrimitiveSchemaDefinitionSchema = exports.EnumSchemaSchema = exports.MultiSelectEnumSchemaSchema = exports.TitledMultiSelectEnumSchemaSchema = exports.UntitledMultiSelectEnumSchemaSchema = exports.SingleSelectEnumSchemaSchema = exports.LegacyTitledEnumSchemaSchema = exports.TitledSingleSelectEnumSchemaSchema = exports.UntitledSingleSelectEnumSchemaSchema = exports.NumberSchemaSchema = exports.StringSchemaSchema = exports.BooleanSchemaSchema = exports.CreateMessageResultWithToolsSchema = exports.CreateMessageResultSchema = exports.CreateMessageRequestSchema = exports.CreateMessageRequestParamsSchema = exports.SamplingMessageSchema = exports.SamplingMessageContentBlockSchema = exports.SamplingContentSchema = exports.ToolResultContentSchema = exports.ToolChoiceSchema = exports.ModelPreferencesSchema = exports.ModelHintSchema = exports.LoggingMessageNotificationSchema = exports.LoggingMessageNotificationParamsSchema = exports.SetLevelRequestSchema = exports.SetLevelRequestParamsSchema = exports.LoggingLevelSchema = exports.ListChangedOptionsBaseSchema = exports.ToolListChangedNotificationSchema = exports.CallToolRequestSchema = exports.CallToolRequestParamsSchema = exports.CompatibilityCallToolResultSchema = exports.CallToolResultSchema = exports.ListToolsResultSchema = void 0;
exports.UrlElicitationRequiredError = exports.McpError = exports.ServerResultSchema = exports.ServerNotificationSchema = exports.ServerRequestSchema = exports.ClientResultSchema = exports.ClientNotificationSchema = exports.ClientRequestSchema = exports.RootsListChangedNotificationSchema = exports.ListRootsResultSchema = void 0;
exports.assertCompleteRequestPrompt = assertCompleteRequestPrompt;
exports.assertCompleteRequestResourceTemplate = assertCompleteRequestResourceTemplate;
const v4_1 = __importDefault(require("zod/v4"));
exports.LATEST_PROTOCOL_VERSION = '2025-11-25';
exports.DEFAULT_NEGOTIATED_PROTOCOL_VERSION = '2025-03-26';
exports.SUPPORTED_PROTOCOL_VERSIONS = [exports.LATEST_PROTOCOL_VERSION, '2025-06-18', '2025-03-26', '2024-11-05', '2024-10-07'];
exports.RELATED_TASK_META_KEY = 'io.modelcontextprotocol/related-task';
/* JSON-RPC types */
exports.JSONRPC_VERSION = '2.0';
/**
 * Assert 'object' type schema.
 *
 * @internal
 */
const AssertObjectSchema = v4_1.default.custom((v) => v !== null && (typeof v === 'object' || typeof v === 'function'));
/**
 * A progress token, used to associate progress notifications with the original request.
 */
exports.ProgressTokenSchema = v4_1.default.union([v4_1.default.string(), v4_1.default.number().int()]);
/**
 * An opaque token used to represent a cursor for pagination.
 */
exports.CursorSchema = v4_1.default.string();
/**
 * Task creation parameters, used to ask that the server create a task to represent a request.
 */
exports.TaskCreationParamsSchema = v4_1.default.looseObject({
    /**
     * Time in milliseconds to keep task results available after completion.
     * If null, the task has unlimited lifetime until manually cleaned up.
     */
    ttl: v4_1.default.union([v4_1.default.number(), v4_1.default.null()]).optional(),
    /**
     * Time in milliseconds to wait between task status requests.
     */
    pollInterval: v4_1.default.number().optional()
});
/**
 * Task association metadata, used to signal which task a message originated from.
 */
exports.RelatedTaskMetadataSchema = v4_1.default.looseObject({
    taskId: v4_1.default.string()
});
const RequestMetaSchema = v4_1.default.looseObject({
    /**
     * If specified, the caller is requesting out-of-band progress notifications for this request (as represented by notifications/progress). The value of this parameter is an opaque token that will be attached to any subsequent notifications. The receiver is not obligated to provide these notifications.
     */
    progressToken: exports.ProgressTokenSchema.optional(),
    /**
     * If specified, this request is related to the provided task.
     */
    [exports.RELATED_TASK_META_KEY]: exports.RelatedTaskMetadataSchema.optional()
});
/**
 * Common params for any request.
 */
const BaseRequestParamsSchema = v4_1.default.looseObject({
    /**
     * If specified, the caller is requesting that the receiver create a task to represent the request.
     * Task creation parameters are now at the top level instead of in _meta.
     */
    task: exports.TaskCreationParamsSchema.optional(),
    /**
     * See [General fields: `_meta`](/specification/draft/basic/index#meta) for notes on `_meta` usage.
     */
    _meta: RequestMetaSchema.optional()
});
exports.RequestSchema = v4_1.default.object({
    method: v4_1.default.string(),
    params: BaseRequestParamsSchema.optional()
});
const NotificationsParamsSchema = v4_1.default.looseObject({
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default
        .object({
        /**
         * If specified, this notification is related to the provided task.
         */
        [exports.RELATED_TASK_META_KEY]: v4_1.default.optional(exports.RelatedTaskMetadataSchema)
    })
        .passthrough()
        .optional()
});
exports.NotificationSchema = v4_1.default.object({
    method: v4_1.default.string(),
    params: NotificationsParamsSchema.optional()
});
exports.ResultSchema = v4_1.default.looseObject({
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default
        .looseObject({
        /**
         * If specified, this result is related to the provided task.
         */
        [exports.RELATED_TASK_META_KEY]: exports.RelatedTaskMetadataSchema.optional()
    })
        .optional()
});
/**
 * A uniquely identifying ID for a request in JSON-RPC.
 */
exports.RequestIdSchema = v4_1.default.union([v4_1.default.string(), v4_1.default.number().int()]);
/**
 * A request that expects a response.
 */
exports.JSONRPCRequestSchema = v4_1.default
    .object({
    jsonrpc: v4_1.default.literal(exports.JSONRPC_VERSION),
    id: exports.RequestIdSchema,
    ...exports.RequestSchema.shape
})
    .strict();
const isJSONRPCRequest = (value) => exports.JSONRPCRequestSchema.safeParse(value).success;
exports.isJSONRPCRequest = isJSONRPCRequest;
/**
 * A notification which does not expect a response.
 */
exports.JSONRPCNotificationSchema = v4_1.default
    .object({
    jsonrpc: v4_1.default.literal(exports.JSONRPC_VERSION),
    ...exports.NotificationSchema.shape
})
    .strict();
const isJSONRPCNotification = (value) => exports.JSONRPCNotificationSchema.safeParse(value).success;
exports.isJSONRPCNotification = isJSONRPCNotification;
/**
 * A successful (non-error) response to a request.
 */
exports.JSONRPCResponseSchema = v4_1.default
    .object({
    jsonrpc: v4_1.default.literal(exports.JSONRPC_VERSION),
    id: exports.RequestIdSchema,
    result: exports.ResultSchema
})
    .strict();
const isJSONRPCResponse = (value) => exports.JSONRPCResponseSchema.safeParse(value).success;
exports.isJSONRPCResponse = isJSONRPCResponse;
/**
 * Error codes defined by the JSON-RPC specification.
 */
var ErrorCode;
(function (ErrorCode) {
    // SDK error codes
    ErrorCode[ErrorCode["ConnectionClosed"] = -32000] = "ConnectionClosed";
    ErrorCode[ErrorCode["RequestTimeout"] = -32001] = "RequestTimeout";
    // Standard JSON-RPC error codes
    ErrorCode[ErrorCode["ParseError"] = -32700] = "ParseError";
    ErrorCode[ErrorCode["InvalidRequest"] = -32600] = "InvalidRequest";
    ErrorCode[ErrorCode["MethodNotFound"] = -32601] = "MethodNotFound";
    ErrorCode[ErrorCode["InvalidParams"] = -32602] = "InvalidParams";
    ErrorCode[ErrorCode["InternalError"] = -32603] = "InternalError";
    // MCP-specific error codes
    ErrorCode[ErrorCode["UrlElicitationRequired"] = -32042] = "UrlElicitationRequired";
})(ErrorCode || (exports.ErrorCode = ErrorCode = {}));
/**
 * A response to a request that indicates an error occurred.
 */
exports.JSONRPCErrorSchema = v4_1.default
    .object({
    jsonrpc: v4_1.default.literal(exports.JSONRPC_VERSION),
    id: exports.RequestIdSchema,
    error: v4_1.default.object({
        /**
         * The error type that occurred.
         */
        code: v4_1.default.number().int(),
        /**
         * A short description of the error. The message SHOULD be limited to a concise single sentence.
         */
        message: v4_1.default.string(),
        /**
         * Additional information about the error. The value of this member is defined by the sender (e.g. detailed error information, nested errors etc.).
         */
        data: v4_1.default.optional(v4_1.default.unknown())
    })
})
    .strict();
const isJSONRPCError = (value) => exports.JSONRPCErrorSchema.safeParse(value).success;
exports.isJSONRPCError = isJSONRPCError;
exports.JSONRPCMessageSchema = v4_1.default.union([exports.JSONRPCRequestSchema, exports.JSONRPCNotificationSchema, exports.JSONRPCResponseSchema, exports.JSONRPCErrorSchema]);
/* Empty result */
/**
 * A response that indicates success but carries no data.
 */
exports.EmptyResultSchema = exports.ResultSchema.strict();
exports.CancelledNotificationParamsSchema = NotificationsParamsSchema.extend({
    /**
     * The ID of the request to cancel.
     *
     * This MUST correspond to the ID of a request previously issued in the same direction.
     */
    requestId: exports.RequestIdSchema,
    /**
     * An optional string describing the reason for the cancellation. This MAY be logged or presented to the user.
     */
    reason: v4_1.default.string().optional()
});
/* Cancellation */
/**
 * This notification can be sent by either side to indicate that it is cancelling a previously-issued request.
 *
 * The request SHOULD still be in-flight, but due to communication latency, it is always possible that this notification MAY arrive after the request has already finished.
 *
 * This notification indicates that the result will be unused, so any associated processing SHOULD cease.
 *
 * A client MUST NOT attempt to cancel its `initialize` request.
 */
exports.CancelledNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/cancelled'),
    params: exports.CancelledNotificationParamsSchema
});
/* Base Metadata */
/**
 * Icon schema for use in tools, prompts, resources, and implementations.
 */
exports.IconSchema = v4_1.default.object({
    /**
     * URL or data URI for the icon.
     */
    src: v4_1.default.string(),
    /**
     * Optional MIME type for the icon.
     */
    mimeType: v4_1.default.string().optional(),
    /**
     * Optional array of strings that specify sizes at which the icon can be used.
     * Each string should be in WxH format (e.g., `"48x48"`, `"96x96"`) or `"any"` for scalable formats like SVG.
     *
     * If not provided, the client should assume that the icon can be used at any size.
     */
    sizes: v4_1.default.array(v4_1.default.string()).optional()
});
/**
 * Base schema to add `icons` property.
 *
 */
exports.IconsSchema = v4_1.default.object({
    /**
     * Optional set of sized icons that the client can display in a user interface.
     *
     * Clients that support rendering icons MUST support at least the following MIME types:
     * - `image/png` - PNG images (safe, universal compatibility)
     * - `image/jpeg` (and `image/jpg`) - JPEG images (safe, universal compatibility)
     *
     * Clients that support rendering icons SHOULD also support:
     * - `image/svg+xml` - SVG images (scalable but requires security precautions)
     * - `image/webp` - WebP images (modern, efficient format)
     */
    icons: v4_1.default.array(exports.IconSchema).optional()
});
/**
 * Base metadata interface for common properties across resources, tools, prompts, and implementations.
 */
exports.BaseMetadataSchema = v4_1.default.object({
    /** Intended for programmatic or logical use, but used as a display name in past specs or fallback */
    name: v4_1.default.string(),
    /**
     * Intended for UI and end-user contexts — optimized to be human-readable and easily understood,
     * even by those unfamiliar with domain-specific terminology.
     *
     * If not provided, the name should be used for display (except for Tool,
     * where `annotations.title` should be given precedence over using `name`,
     * if present).
     */
    title: v4_1.default.string().optional()
});
/* Initialization */
/**
 * Describes the name and version of an MCP implementation.
 */
exports.ImplementationSchema = exports.BaseMetadataSchema.extend({
    ...exports.BaseMetadataSchema.shape,
    ...exports.IconsSchema.shape,
    version: v4_1.default.string(),
    /**
     * An optional URL of the website for this implementation.
     */
    websiteUrl: v4_1.default.string().optional()
});
const FormElicitationCapabilitySchema = v4_1.default.intersection(v4_1.default.object({
    applyDefaults: v4_1.default.boolean().optional()
}), v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()));
const ElicitationCapabilitySchema = v4_1.default.preprocess(value => {
    if (value && typeof value === 'object' && !Array.isArray(value)) {
        if (Object.keys(value).length === 0) {
            return { form: {} };
        }
    }
    return value;
}, v4_1.default.intersection(v4_1.default.object({
    form: FormElicitationCapabilitySchema.optional(),
    url: AssertObjectSchema.optional()
}), v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()));
/**
 * Task capabilities for clients, indicating which request types support task creation.
 */
exports.ClientTasksCapabilitySchema = v4_1.default
    .object({
    /**
     * Present if the client supports listing tasks.
     */
    list: v4_1.default.optional(v4_1.default.object({}).passthrough()),
    /**
     * Present if the client supports cancelling tasks.
     */
    cancel: v4_1.default.optional(v4_1.default.object({}).passthrough()),
    /**
     * Capabilities for task creation on specific request types.
     */
    requests: v4_1.default.optional(v4_1.default
        .object({
        /**
         * Task support for sampling requests.
         */
        sampling: v4_1.default.optional(v4_1.default
            .object({
            createMessage: v4_1.default.optional(v4_1.default.object({}).passthrough())
        })
            .passthrough()),
        /**
         * Task support for elicitation requests.
         */
        elicitation: v4_1.default.optional(v4_1.default
            .object({
            create: v4_1.default.optional(v4_1.default.object({}).passthrough())
        })
            .passthrough())
    })
        .passthrough())
})
    .passthrough();
/**
 * Task capabilities for servers, indicating which request types support task creation.
 */
exports.ServerTasksCapabilitySchema = v4_1.default
    .object({
    /**
     * Present if the server supports listing tasks.
     */
    list: v4_1.default.optional(v4_1.default.object({}).passthrough()),
    /**
     * Present if the server supports cancelling tasks.
     */
    cancel: v4_1.default.optional(v4_1.default.object({}).passthrough()),
    /**
     * Capabilities for task creation on specific request types.
     */
    requests: v4_1.default.optional(v4_1.default
        .object({
        /**
         * Task support for tool requests.
         */
        tools: v4_1.default.optional(v4_1.default
            .object({
            call: v4_1.default.optional(v4_1.default.object({}).passthrough())
        })
            .passthrough())
    })
        .passthrough())
})
    .passthrough();
/**
 * Capabilities a client may support. Known capabilities are defined here, in this schema, but this is not a closed set: any client can define its own, additional capabilities.
 */
exports.ClientCapabilitiesSchema = v4_1.default.object({
    /**
     * Experimental, non-standard capabilities that the client supports.
     */
    experimental: v4_1.default.record(v4_1.default.string(), AssertObjectSchema).optional(),
    /**
     * Present if the client supports sampling from an LLM.
     */
    sampling: v4_1.default
        .object({
        /**
         * Present if the client supports context inclusion via includeContext parameter.
         * If not declared, servers SHOULD only use `includeContext: "none"` (or omit it).
         */
        context: AssertObjectSchema.optional(),
        /**
         * Present if the client supports tool use via tools and toolChoice parameters.
         */
        tools: AssertObjectSchema.optional()
    })
        .optional(),
    /**
     * Present if the client supports eliciting user input.
     */
    elicitation: ElicitationCapabilitySchema.optional(),
    /**
     * Present if the client supports listing roots.
     */
    roots: v4_1.default
        .object({
        /**
         * Whether the client supports issuing notifications for changes to the roots list.
         */
        listChanged: v4_1.default.boolean().optional()
    })
        .optional(),
    /**
     * Present if the client supports task creation.
     */
    tasks: v4_1.default.optional(exports.ClientTasksCapabilitySchema)
});
exports.InitializeRequestParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The latest version of the Model Context Protocol that the client supports. The client MAY decide to support older versions as well.
     */
    protocolVersion: v4_1.default.string(),
    capabilities: exports.ClientCapabilitiesSchema,
    clientInfo: exports.ImplementationSchema
});
/**
 * This request is sent from the client to the server when it first connects, asking it to begin initialization.
 */
exports.InitializeRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('initialize'),
    params: exports.InitializeRequestParamsSchema
});
const isInitializeRequest = (value) => exports.InitializeRequestSchema.safeParse(value).success;
exports.isInitializeRequest = isInitializeRequest;
/**
 * Capabilities that a server may support. Known capabilities are defined here, in this schema, but this is not a closed set: any server can define its own, additional capabilities.
 */
exports.ServerCapabilitiesSchema = v4_1.default
    .object({
    /**
     * Experimental, non-standard capabilities that the server supports.
     */
    experimental: v4_1.default.record(v4_1.default.string(), AssertObjectSchema).optional(),
    /**
     * Present if the server supports sending log messages to the client.
     */
    logging: AssertObjectSchema.optional(),
    /**
     * Present if the server supports sending completions to the client.
     */
    completions: AssertObjectSchema.optional(),
    /**
     * Present if the server offers any prompt templates.
     */
    prompts: v4_1.default.optional(v4_1.default.object({
        /**
         * Whether this server supports issuing notifications for changes to the prompt list.
         */
        listChanged: v4_1.default.optional(v4_1.default.boolean())
    })),
    /**
     * Present if the server offers any resources to read.
     */
    resources: v4_1.default
        .object({
        /**
         * Whether this server supports clients subscribing to resource updates.
         */
        subscribe: v4_1.default.boolean().optional(),
        /**
         * Whether this server supports issuing notifications for changes to the resource list.
         */
        listChanged: v4_1.default.boolean().optional()
    })
        .optional(),
    /**
     * Present if the server offers any tools to call.
     */
    tools: v4_1.default
        .object({
        /**
         * Whether this server supports issuing notifications for changes to the tool list.
         */
        listChanged: v4_1.default.boolean().optional()
    })
        .optional(),
    /**
     * Present if the server supports task creation.
     */
    tasks: v4_1.default.optional(exports.ServerTasksCapabilitySchema)
})
    .passthrough();
/**
 * After receiving an initialize request from the client, the server sends this response.
 */
exports.InitializeResultSchema = exports.ResultSchema.extend({
    /**
     * The version of the Model Context Protocol that the server wants to use. This may not match the version that the client requested. If the client cannot support this version, it MUST disconnect.
     */
    protocolVersion: v4_1.default.string(),
    capabilities: exports.ServerCapabilitiesSchema,
    serverInfo: exports.ImplementationSchema,
    /**
     * Instructions describing how to use the server and its features.
     *
     * This can be used by clients to improve the LLM's understanding of available tools, resources, etc. It can be thought of like a "hint" to the model. For example, this information MAY be added to the system prompt.
     */
    instructions: v4_1.default.string().optional()
});
/**
 * This notification is sent from the client to the server after initialization has finished.
 */
exports.InitializedNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/initialized')
});
const isInitializedNotification = (value) => exports.InitializedNotificationSchema.safeParse(value).success;
exports.isInitializedNotification = isInitializedNotification;
/* Ping */
/**
 * A ping, issued by either the server or the client, to check that the other party is still alive. The receiver must promptly respond, or else may be disconnected.
 */
exports.PingRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('ping')
});
/* Progress notifications */
exports.ProgressSchema = v4_1.default.object({
    /**
     * The progress thus far. This should increase every time progress is made, even if the total is unknown.
     */
    progress: v4_1.default.number(),
    /**
     * Total number of items to process (or total progress required), if known.
     */
    total: v4_1.default.optional(v4_1.default.number()),
    /**
     * An optional message describing the current progress.
     */
    message: v4_1.default.optional(v4_1.default.string())
});
exports.ProgressNotificationParamsSchema = v4_1.default.object({
    ...NotificationsParamsSchema.shape,
    ...exports.ProgressSchema.shape,
    /**
     * The progress token which was given in the initial request, used to associate this notification with the request that is proceeding.
     */
    progressToken: exports.ProgressTokenSchema
});
/**
 * An out-of-band notification used to inform the receiver of a progress update for a long-running request.
 *
 * @category notifications/progress
 */
exports.ProgressNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/progress'),
    params: exports.ProgressNotificationParamsSchema
});
exports.PaginatedRequestParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * An opaque token representing the current pagination position.
     * If provided, the server should return results starting after this cursor.
     */
    cursor: exports.CursorSchema.optional()
});
/* Pagination */
exports.PaginatedRequestSchema = exports.RequestSchema.extend({
    params: exports.PaginatedRequestParamsSchema.optional()
});
exports.PaginatedResultSchema = exports.ResultSchema.extend({
    /**
     * An opaque token representing the pagination position after the last returned result.
     * If present, there may be more results available.
     */
    nextCursor: v4_1.default.optional(exports.CursorSchema)
});
/* Tasks */
/**
 * A pollable state object associated with a request.
 */
exports.TaskSchema = v4_1.default.object({
    taskId: v4_1.default.string(),
    status: v4_1.default.enum(['working', 'input_required', 'completed', 'failed', 'cancelled']),
    /**
     * Time in milliseconds to keep task results available after completion.
     * If null, the task has unlimited lifetime until manually cleaned up.
     */
    ttl: v4_1.default.union([v4_1.default.number(), v4_1.default.null()]),
    /**
     * ISO 8601 timestamp when the task was created.
     */
    createdAt: v4_1.default.string(),
    /**
     * ISO 8601 timestamp when the task was last updated.
     */
    lastUpdatedAt: v4_1.default.string(),
    pollInterval: v4_1.default.optional(v4_1.default.number()),
    /**
     * Optional diagnostic message for failed tasks or other status information.
     */
    statusMessage: v4_1.default.optional(v4_1.default.string())
});
/**
 * Result returned when a task is created, containing the task data wrapped in a task field.
 */
exports.CreateTaskResultSchema = exports.ResultSchema.extend({
    task: exports.TaskSchema
});
/**
 * Parameters for task status notification.
 */
exports.TaskStatusNotificationParamsSchema = NotificationsParamsSchema.merge(exports.TaskSchema);
/**
 * A notification sent when a task's status changes.
 */
exports.TaskStatusNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/tasks/status'),
    params: exports.TaskStatusNotificationParamsSchema
});
/**
 * A request to get the state of a specific task.
 */
exports.GetTaskRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('tasks/get'),
    params: BaseRequestParamsSchema.extend({
        taskId: v4_1.default.string()
    })
});
/**
 * The response to a tasks/get request.
 */
exports.GetTaskResultSchema = exports.ResultSchema.merge(exports.TaskSchema);
/**
 * A request to get the result of a specific task.
 */
exports.GetTaskPayloadRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('tasks/result'),
    params: BaseRequestParamsSchema.extend({
        taskId: v4_1.default.string()
    })
});
/**
 * A request to list tasks.
 */
exports.ListTasksRequestSchema = exports.PaginatedRequestSchema.extend({
    method: v4_1.default.literal('tasks/list')
});
/**
 * The response to a tasks/list request.
 */
exports.ListTasksResultSchema = exports.PaginatedResultSchema.extend({
    tasks: v4_1.default.array(exports.TaskSchema)
});
/**
 * A request to cancel a specific task.
 */
exports.CancelTaskRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('tasks/cancel'),
    params: BaseRequestParamsSchema.extend({
        taskId: v4_1.default.string()
    })
});
/**
 * The response to a tasks/cancel request.
 */
exports.CancelTaskResultSchema = exports.ResultSchema.merge(exports.TaskSchema);
/* Resources */
/**
 * The contents of a specific resource or sub-resource.
 */
exports.ResourceContentsSchema = v4_1.default.object({
    /**
     * The URI of this resource.
     */
    uri: v4_1.default.string(),
    /**
     * The MIME type of this resource, if known.
     */
    mimeType: v4_1.default.optional(v4_1.default.string()),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
exports.TextResourceContentsSchema = exports.ResourceContentsSchema.extend({
    /**
     * The text of the item. This must only be set if the item can actually be represented as text (not binary data).
     */
    text: v4_1.default.string()
});
/**
 * A Zod schema for validating Base64 strings that is more performant and
 * robust for very large inputs than the default regex-based check. It avoids
 * stack overflows by using the native `atob` function for validation.
 */
const Base64Schema = v4_1.default.string().refine(val => {
    try {
        // atob throws a DOMException if the string contains characters
        // that are not part of the Base64 character set.
        atob(val);
        return true;
    }
    catch (_a) {
        return false;
    }
}, { message: 'Invalid Base64 string' });
exports.BlobResourceContentsSchema = exports.ResourceContentsSchema.extend({
    /**
     * A base64-encoded string representing the binary data of the item.
     */
    blob: Base64Schema
});
/**
 * The sender or recipient of messages and data in a conversation.
 */
exports.RoleSchema = v4_1.default.enum(['user', 'assistant']);
/**
 * Optional annotations providing clients additional context about a resource.
 */
exports.AnnotationsSchema = v4_1.default.object({
    /**
     * Intended audience(s) for the resource.
     */
    audience: v4_1.default.array(exports.RoleSchema).optional(),
    /**
     * Importance hint for the resource, from 0 (least) to 1 (most).
     */
    priority: v4_1.default.number().min(0).max(1).optional(),
    /**
     * ISO 8601 timestamp for the most recent modification.
     */
    lastModified: v4_1.default.iso.datetime({ offset: true }).optional()
});
/**
 * A known resource that the server is capable of reading.
 */
exports.ResourceSchema = v4_1.default.object({
    ...exports.BaseMetadataSchema.shape,
    ...exports.IconsSchema.shape,
    /**
     * The URI of this resource.
     */
    uri: v4_1.default.string(),
    /**
     * A description of what this resource represents.
     *
     * This can be used by clients to improve the LLM's understanding of available resources. It can be thought of like a "hint" to the model.
     */
    description: v4_1.default.optional(v4_1.default.string()),
    /**
     * The MIME type of this resource, if known.
     */
    mimeType: v4_1.default.optional(v4_1.default.string()),
    /**
     * Optional annotations for the client.
     */
    annotations: exports.AnnotationsSchema.optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.optional(v4_1.default.looseObject({}))
});
/**
 * A template description for resources available on the server.
 */
exports.ResourceTemplateSchema = v4_1.default.object({
    ...exports.BaseMetadataSchema.shape,
    ...exports.IconsSchema.shape,
    /**
     * A URI template (according to RFC 6570) that can be used to construct resource URIs.
     */
    uriTemplate: v4_1.default.string(),
    /**
     * A description of what this template is for.
     *
     * This can be used by clients to improve the LLM's understanding of available resources. It can be thought of like a "hint" to the model.
     */
    description: v4_1.default.optional(v4_1.default.string()),
    /**
     * The MIME type for all resources that match this template. This should only be included if all resources matching this template have the same type.
     */
    mimeType: v4_1.default.optional(v4_1.default.string()),
    /**
     * Optional annotations for the client.
     */
    annotations: exports.AnnotationsSchema.optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.optional(v4_1.default.looseObject({}))
});
/**
 * Sent from the client to request a list of resources the server has.
 */
exports.ListResourcesRequestSchema = exports.PaginatedRequestSchema.extend({
    method: v4_1.default.literal('resources/list')
});
/**
 * The server's response to a resources/list request from the client.
 */
exports.ListResourcesResultSchema = exports.PaginatedResultSchema.extend({
    resources: v4_1.default.array(exports.ResourceSchema)
});
/**
 * Sent from the client to request a list of resource templates the server has.
 */
exports.ListResourceTemplatesRequestSchema = exports.PaginatedRequestSchema.extend({
    method: v4_1.default.literal('resources/templates/list')
});
/**
 * The server's response to a resources/templates/list request from the client.
 */
exports.ListResourceTemplatesResultSchema = exports.PaginatedResultSchema.extend({
    resourceTemplates: v4_1.default.array(exports.ResourceTemplateSchema)
});
exports.ResourceRequestParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The URI of the resource to read. The URI can use any protocol; it is up to the server how to interpret it.
     *
     * @format uri
     */
    uri: v4_1.default.string()
});
/**
 * Parameters for a `resources/read` request.
 */
exports.ReadResourceRequestParamsSchema = exports.ResourceRequestParamsSchema;
/**
 * Sent from the client to the server, to read a specific resource URI.
 */
exports.ReadResourceRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('resources/read'),
    params: exports.ReadResourceRequestParamsSchema
});
/**
 * The server's response to a resources/read request from the client.
 */
exports.ReadResourceResultSchema = exports.ResultSchema.extend({
    contents: v4_1.default.array(v4_1.default.union([exports.TextResourceContentsSchema, exports.BlobResourceContentsSchema]))
});
/**
 * An optional notification from the server to the client, informing it that the list of resources it can read from has changed. This may be issued by servers without any previous subscription from the client.
 */
exports.ResourceListChangedNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/resources/list_changed')
});
exports.SubscribeRequestParamsSchema = exports.ResourceRequestParamsSchema;
/**
 * Sent from the client to request resources/updated notifications from the server whenever a particular resource changes.
 */
exports.SubscribeRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('resources/subscribe'),
    params: exports.SubscribeRequestParamsSchema
});
exports.UnsubscribeRequestParamsSchema = exports.ResourceRequestParamsSchema;
/**
 * Sent from the client to request cancellation of resources/updated notifications from the server. This should follow a previous resources/subscribe request.
 */
exports.UnsubscribeRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('resources/unsubscribe'),
    params: exports.UnsubscribeRequestParamsSchema
});
/**
 * Parameters for a `notifications/resources/updated` notification.
 */
exports.ResourceUpdatedNotificationParamsSchema = NotificationsParamsSchema.extend({
    /**
     * The URI of the resource that has been updated. This might be a sub-resource of the one that the client actually subscribed to.
     */
    uri: v4_1.default.string()
});
/**
 * A notification from the server to the client, informing it that a resource has changed and may need to be read again. This should only be sent if the client previously sent a resources/subscribe request.
 */
exports.ResourceUpdatedNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/resources/updated'),
    params: exports.ResourceUpdatedNotificationParamsSchema
});
/* Prompts */
/**
 * Describes an argument that a prompt can accept.
 */
exports.PromptArgumentSchema = v4_1.default.object({
    /**
     * The name of the argument.
     */
    name: v4_1.default.string(),
    /**
     * A human-readable description of the argument.
     */
    description: v4_1.default.optional(v4_1.default.string()),
    /**
     * Whether this argument must be provided.
     */
    required: v4_1.default.optional(v4_1.default.boolean())
});
/**
 * A prompt or prompt template that the server offers.
 */
exports.PromptSchema = v4_1.default.object({
    ...exports.BaseMetadataSchema.shape,
    ...exports.IconsSchema.shape,
    /**
     * An optional description of what this prompt provides
     */
    description: v4_1.default.optional(v4_1.default.string()),
    /**
     * A list of arguments to use for templating the prompt.
     */
    arguments: v4_1.default.optional(v4_1.default.array(exports.PromptArgumentSchema)),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.optional(v4_1.default.looseObject({}))
});
/**
 * Sent from the client to request a list of prompts and prompt templates the server has.
 */
exports.ListPromptsRequestSchema = exports.PaginatedRequestSchema.extend({
    method: v4_1.default.literal('prompts/list')
});
/**
 * The server's response to a prompts/list request from the client.
 */
exports.ListPromptsResultSchema = exports.PaginatedResultSchema.extend({
    prompts: v4_1.default.array(exports.PromptSchema)
});
/**
 * Parameters for a `prompts/get` request.
 */
exports.GetPromptRequestParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The name of the prompt or prompt template.
     */
    name: v4_1.default.string(),
    /**
     * Arguments to use for templating the prompt.
     */
    arguments: v4_1.default.record(v4_1.default.string(), v4_1.default.string()).optional()
});
/**
 * Used by the client to get a prompt provided by the server.
 */
exports.GetPromptRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('prompts/get'),
    params: exports.GetPromptRequestParamsSchema
});
/**
 * Text provided to or from an LLM.
 */
exports.TextContentSchema = v4_1.default.object({
    type: v4_1.default.literal('text'),
    /**
     * The text content of the message.
     */
    text: v4_1.default.string(),
    /**
     * Optional annotations for the client.
     */
    annotations: exports.AnnotationsSchema.optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
/**
 * An image provided to or from an LLM.
 */
exports.ImageContentSchema = v4_1.default.object({
    type: v4_1.default.literal('image'),
    /**
     * The base64-encoded image data.
     */
    data: Base64Schema,
    /**
     * The MIME type of the image. Different providers may support different image types.
     */
    mimeType: v4_1.default.string(),
    /**
     * Optional annotations for the client.
     */
    annotations: exports.AnnotationsSchema.optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
/**
 * An Audio provided to or from an LLM.
 */
exports.AudioContentSchema = v4_1.default.object({
    type: v4_1.default.literal('audio'),
    /**
     * The base64-encoded audio data.
     */
    data: Base64Schema,
    /**
     * The MIME type of the audio. Different providers may support different audio types.
     */
    mimeType: v4_1.default.string(),
    /**
     * Optional annotations for the client.
     */
    annotations: exports.AnnotationsSchema.optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
/**
 * A tool call request from an assistant (LLM).
 * Represents the assistant's request to use a tool.
 */
exports.ToolUseContentSchema = v4_1.default
    .object({
    type: v4_1.default.literal('tool_use'),
    /**
     * The name of the tool to invoke.
     * Must match a tool name from the request's tools array.
     */
    name: v4_1.default.string(),
    /**
     * Unique identifier for this tool call.
     * Used to correlate with ToolResultContent in subsequent messages.
     */
    id: v4_1.default.string(),
    /**
     * Arguments to pass to the tool.
     * Must conform to the tool's inputSchema.
     */
    input: v4_1.default.object({}).passthrough(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.optional(v4_1.default.object({}).passthrough())
})
    .passthrough();
/**
 * The contents of a resource, embedded into a prompt or tool call result.
 */
exports.EmbeddedResourceSchema = v4_1.default.object({
    type: v4_1.default.literal('resource'),
    resource: v4_1.default.union([exports.TextResourceContentsSchema, exports.BlobResourceContentsSchema]),
    /**
     * Optional annotations for the client.
     */
    annotations: exports.AnnotationsSchema.optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
/**
 * A resource that the server is capable of reading, included in a prompt or tool call result.
 *
 * Note: resource links returned by tools are not guaranteed to appear in the results of `resources/list` requests.
 */
exports.ResourceLinkSchema = exports.ResourceSchema.extend({
    type: v4_1.default.literal('resource_link')
});
/**
 * A content block that can be used in prompts and tool results.
 */
exports.ContentBlockSchema = v4_1.default.union([
    exports.TextContentSchema,
    exports.ImageContentSchema,
    exports.AudioContentSchema,
    exports.ResourceLinkSchema,
    exports.EmbeddedResourceSchema
]);
/**
 * Describes a message returned as part of a prompt.
 */
exports.PromptMessageSchema = v4_1.default.object({
    role: exports.RoleSchema,
    content: exports.ContentBlockSchema
});
/**
 * The server's response to a prompts/get request from the client.
 */
exports.GetPromptResultSchema = exports.ResultSchema.extend({
    /**
     * An optional description for the prompt.
     */
    description: v4_1.default.optional(v4_1.default.string()),
    messages: v4_1.default.array(exports.PromptMessageSchema)
});
/**
 * An optional notification from the server to the client, informing it that the list of prompts it offers has changed. This may be issued by servers without any previous subscription from the client.
 */
exports.PromptListChangedNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/prompts/list_changed')
});
/* Tools */
/**
 * Additional properties describing a Tool to clients.
 *
 * NOTE: all properties in ToolAnnotations are **hints**.
 * They are not guaranteed to provide a faithful description of
 * tool behavior (including descriptive properties like `title`).
 *
 * Clients should never make tool use decisions based on ToolAnnotations
 * received from untrusted servers.
 */
exports.ToolAnnotationsSchema = v4_1.default.object({
    /**
     * A human-readable title for the tool.
     */
    title: v4_1.default.string().optional(),
    /**
     * If true, the tool does not modify its environment.
     *
     * Default: false
     */
    readOnlyHint: v4_1.default.boolean().optional(),
    /**
     * If true, the tool may perform destructive updates to its environment.
     * If false, the tool performs only additive updates.
     *
     * (This property is meaningful only when `readOnlyHint == false`)
     *
     * Default: true
     */
    destructiveHint: v4_1.default.boolean().optional(),
    /**
     * If true, calling the tool repeatedly with the same arguments
     * will have no additional effect on the its environment.
     *
     * (This property is meaningful only when `readOnlyHint == false`)
     *
     * Default: false
     */
    idempotentHint: v4_1.default.boolean().optional(),
    /**
     * If true, this tool may interact with an "open world" of external
     * entities. If false, the tool's domain of interaction is closed.
     * For example, the world of a web search tool is open, whereas that
     * of a memory tool is not.
     *
     * Default: true
     */
    openWorldHint: v4_1.default.boolean().optional()
});
/**
 * Execution-related properties for a tool.
 */
exports.ToolExecutionSchema = v4_1.default.object({
    /**
     * Indicates the tool's preference for task-augmented execution.
     * - "required": Clients MUST invoke the tool as a task
     * - "optional": Clients MAY invoke the tool as a task or normal request
     * - "forbidden": Clients MUST NOT attempt to invoke the tool as a task
     *
     * If not present, defaults to "forbidden".
     */
    taskSupport: v4_1.default.enum(['required', 'optional', 'forbidden']).optional()
});
/**
 * Definition for a tool the client can call.
 */
exports.ToolSchema = v4_1.default.object({
    ...exports.BaseMetadataSchema.shape,
    ...exports.IconsSchema.shape,
    /**
     * A human-readable description of the tool.
     */
    description: v4_1.default.string().optional(),
    /**
     * A JSON Schema 2020-12 object defining the expected parameters for the tool.
     * Must have type: 'object' at the root level per MCP spec.
     */
    inputSchema: v4_1.default
        .object({
        type: v4_1.default.literal('object'),
        properties: v4_1.default.record(v4_1.default.string(), AssertObjectSchema).optional(),
        required: v4_1.default.array(v4_1.default.string()).optional()
    })
        .catchall(v4_1.default.unknown()),
    /**
     * An optional JSON Schema 2020-12 object defining the structure of the tool's output
     * returned in the structuredContent field of a CallToolResult.
     * Must have type: 'object' at the root level per MCP spec.
     */
    outputSchema: v4_1.default
        .object({
        type: v4_1.default.literal('object'),
        properties: v4_1.default.record(v4_1.default.string(), AssertObjectSchema).optional(),
        required: v4_1.default.array(v4_1.default.string()).optional()
    })
        .catchall(v4_1.default.unknown())
        .optional(),
    /**
     * Optional additional tool information.
     */
    annotations: v4_1.default.optional(exports.ToolAnnotationsSchema),
    /**
     * Execution-related properties for this tool.
     */
    execution: v4_1.default.optional(exports.ToolExecutionSchema),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
/**
 * Sent from the client to request a list of tools the server has.
 */
exports.ListToolsRequestSchema = exports.PaginatedRequestSchema.extend({
    method: v4_1.default.literal('tools/list')
});
/**
 * The server's response to a tools/list request from the client.
 */
exports.ListToolsResultSchema = exports.PaginatedResultSchema.extend({
    tools: v4_1.default.array(exports.ToolSchema)
});
/**
 * The server's response to a tool call.
 */
exports.CallToolResultSchema = exports.ResultSchema.extend({
    /**
     * A list of content objects that represent the result of the tool call.
     *
     * If the Tool does not define an outputSchema, this field MUST be present in the result.
     * For backwards compatibility, this field is always present, but it may be empty.
     */
    content: v4_1.default.array(exports.ContentBlockSchema).default([]),
    /**
     * An object containing structured tool output.
     *
     * If the Tool defines an outputSchema, this field MUST be present in the result, and contain a JSON object that matches the schema.
     */
    structuredContent: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional(),
    /**
     * Whether the tool call ended in an error.
     *
     * If not set, this is assumed to be false (the call was successful).
     *
     * Any errors that originate from the tool SHOULD be reported inside the result
     * object, with `isError` set to true, _not_ as an MCP protocol-level error
     * response. Otherwise, the LLM would not be able to see that an error occurred
     * and self-correct.
     *
     * However, any errors in _finding_ the tool, an error indicating that the
     * server does not support tool calls, or any other exceptional conditions,
     * should be reported as an MCP error response.
     */
    isError: v4_1.default.optional(v4_1.default.boolean())
});
/**
 * CallToolResultSchema extended with backwards compatibility to protocol version 2024-10-07.
 */
exports.CompatibilityCallToolResultSchema = exports.CallToolResultSchema.or(exports.ResultSchema.extend({
    toolResult: v4_1.default.unknown()
}));
/**
 * Parameters for a `tools/call` request.
 */
exports.CallToolRequestParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The name of the tool to call.
     */
    name: v4_1.default.string(),
    /**
     * Arguments to pass to the tool.
     */
    arguments: v4_1.default.optional(v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()))
});
/**
 * Used by the client to invoke a tool provided by the server.
 */
exports.CallToolRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('tools/call'),
    params: exports.CallToolRequestParamsSchema
});
/**
 * An optional notification from the server to the client, informing it that the list of tools it offers has changed. This may be issued by servers without any previous subscription from the client.
 */
exports.ToolListChangedNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/tools/list_changed')
});
/**
 * Base schema for list changed subscription options (without callback).
 * Used internally for Zod validation of autoRefresh and debounceMs.
 */
exports.ListChangedOptionsBaseSchema = v4_1.default.object({
    /**
     * If true, the list will be refreshed automatically when a list changed notification is received.
     * The callback will be called with the updated list.
     *
     * If false, the callback will be called with null items, allowing manual refresh.
     *
     * @default true
     */
    autoRefresh: v4_1.default.boolean().default(true),
    /**
     * Debounce time in milliseconds for list changed notification processing.
     *
     * Multiple notifications received within this timeframe will only trigger one refresh.
     * Set to 0 to disable debouncing.
     *
     * @default 300
     */
    debounceMs: v4_1.default.number().int().nonnegative().default(300)
});
/* Logging */
/**
 * The severity of a log message.
 */
exports.LoggingLevelSchema = v4_1.default.enum(['debug', 'info', 'notice', 'warning', 'error', 'critical', 'alert', 'emergency']);
/**
 * Parameters for a `logging/setLevel` request.
 */
exports.SetLevelRequestParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The level of logging that the client wants to receive from the server. The server should send all logs at this level and higher (i.e., more severe) to the client as notifications/logging/message.
     */
    level: exports.LoggingLevelSchema
});
/**
 * A request from the client to the server, to enable or adjust logging.
 */
exports.SetLevelRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('logging/setLevel'),
    params: exports.SetLevelRequestParamsSchema
});
/**
 * Parameters for a `notifications/message` notification.
 */
exports.LoggingMessageNotificationParamsSchema = NotificationsParamsSchema.extend({
    /**
     * The severity of this log message.
     */
    level: exports.LoggingLevelSchema,
    /**
     * An optional name of the logger issuing this message.
     */
    logger: v4_1.default.string().optional(),
    /**
     * The data to be logged, such as a string message or an object. Any JSON serializable type is allowed here.
     */
    data: v4_1.default.unknown()
});
/**
 * Notification of a log message passed from server to client. If no logging/setLevel request has been sent from the client, the server MAY decide which messages to send automatically.
 */
exports.LoggingMessageNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/message'),
    params: exports.LoggingMessageNotificationParamsSchema
});
/* Sampling */
/**
 * Hints to use for model selection.
 */
exports.ModelHintSchema = v4_1.default.object({
    /**
     * A hint for a model name.
     */
    name: v4_1.default.string().optional()
});
/**
 * The server's preferences for model selection, requested of the client during sampling.
 */
exports.ModelPreferencesSchema = v4_1.default.object({
    /**
     * Optional hints to use for model selection.
     */
    hints: v4_1.default.optional(v4_1.default.array(exports.ModelHintSchema)),
    /**
     * How much to prioritize cost when selecting a model.
     */
    costPriority: v4_1.default.optional(v4_1.default.number().min(0).max(1)),
    /**
     * How much to prioritize sampling speed (latency) when selecting a model.
     */
    speedPriority: v4_1.default.optional(v4_1.default.number().min(0).max(1)),
    /**
     * How much to prioritize intelligence and capabilities when selecting a model.
     */
    intelligencePriority: v4_1.default.optional(v4_1.default.number().min(0).max(1))
});
/**
 * Controls tool usage behavior in sampling requests.
 */
exports.ToolChoiceSchema = v4_1.default.object({
    /**
     * Controls when tools are used:
     * - "auto": Model decides whether to use tools (default)
     * - "required": Model MUST use at least one tool before completing
     * - "none": Model MUST NOT use any tools
     */
    mode: v4_1.default.optional(v4_1.default.enum(['auto', 'required', 'none']))
});
/**
 * The result of a tool execution, provided by the user (server).
 * Represents the outcome of invoking a tool requested via ToolUseContent.
 */
exports.ToolResultContentSchema = v4_1.default
    .object({
    type: v4_1.default.literal('tool_result'),
    toolUseId: v4_1.default.string().describe('The unique identifier for the corresponding tool call.'),
    content: v4_1.default.array(exports.ContentBlockSchema).default([]),
    structuredContent: v4_1.default.object({}).passthrough().optional(),
    isError: v4_1.default.optional(v4_1.default.boolean()),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.optional(v4_1.default.object({}).passthrough())
})
    .passthrough();
/**
 * Basic content types for sampling responses (without tool use).
 * Used for backwards-compatible CreateMessageResult when tools are not used.
 */
exports.SamplingContentSchema = v4_1.default.discriminatedUnion('type', [exports.TextContentSchema, exports.ImageContentSchema, exports.AudioContentSchema]);
/**
 * Content block types allowed in sampling messages.
 * This includes text, image, audio, tool use requests, and tool results.
 */
exports.SamplingMessageContentBlockSchema = v4_1.default.discriminatedUnion('type', [
    exports.TextContentSchema,
    exports.ImageContentSchema,
    exports.AudioContentSchema,
    exports.ToolUseContentSchema,
    exports.ToolResultContentSchema
]);
/**
 * Describes a message issued to or received from an LLM API.
 */
exports.SamplingMessageSchema = v4_1.default
    .object({
    role: exports.RoleSchema,
    content: v4_1.default.union([exports.SamplingMessageContentBlockSchema, v4_1.default.array(exports.SamplingMessageContentBlockSchema)]),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.optional(v4_1.default.object({}).passthrough())
})
    .passthrough();
/**
 * Parameters for a `sampling/createMessage` request.
 */
exports.CreateMessageRequestParamsSchema = BaseRequestParamsSchema.extend({
    messages: v4_1.default.array(exports.SamplingMessageSchema),
    /**
     * The server's preferences for which model to select. The client MAY modify or omit this request.
     */
    modelPreferences: exports.ModelPreferencesSchema.optional(),
    /**
     * An optional system prompt the server wants to use for sampling. The client MAY modify or omit this prompt.
     */
    systemPrompt: v4_1.default.string().optional(),
    /**
     * A request to include context from one or more MCP servers (including the caller), to be attached to the prompt.
     * The client MAY ignore this request.
     *
     * Default is "none". Values "thisServer" and "allServers" are soft-deprecated. Servers SHOULD only use these values if the client
     * declares ClientCapabilities.sampling.context. These values may be removed in future spec releases.
     */
    includeContext: v4_1.default.enum(['none', 'thisServer', 'allServers']).optional(),
    temperature: v4_1.default.number().optional(),
    /**
     * The requested maximum number of tokens to sample (to prevent runaway completions).
     *
     * The client MAY choose to sample fewer tokens than the requested maximum.
     */
    maxTokens: v4_1.default.number().int(),
    stopSequences: v4_1.default.array(v4_1.default.string()).optional(),
    /**
     * Optional metadata to pass through to the LLM provider. The format of this metadata is provider-specific.
     */
    metadata: AssertObjectSchema.optional(),
    /**
     * Tools that the model may use during generation.
     * The client MUST return an error if this field is provided but ClientCapabilities.sampling.tools is not declared.
     */
    tools: v4_1.default.optional(v4_1.default.array(exports.ToolSchema)),
    /**
     * Controls how the model uses tools.
     * The client MUST return an error if this field is provided but ClientCapabilities.sampling.tools is not declared.
     * Default is `{ mode: "auto" }`.
     */
    toolChoice: v4_1.default.optional(exports.ToolChoiceSchema)
});
/**
 * A request from the server to sample an LLM via the client. The client has full discretion over which model to select. The client should also inform the user before beginning sampling, to allow them to inspect the request (human in the loop) and decide whether to approve it.
 */
exports.CreateMessageRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('sampling/createMessage'),
    params: exports.CreateMessageRequestParamsSchema
});
/**
 * The client's response to a sampling/create_message request from the server.
 * This is the backwards-compatible version that returns single content (no arrays).
 * Used when the request does not include tools.
 */
exports.CreateMessageResultSchema = exports.ResultSchema.extend({
    /**
     * The name of the model that generated the message.
     */
    model: v4_1.default.string(),
    /**
     * The reason why sampling stopped, if known.
     *
     * Standard values:
     * - "endTurn": Natural end of the assistant's turn
     * - "stopSequence": A stop sequence was encountered
     * - "maxTokens": Maximum token limit was reached
     *
     * This field is an open string to allow for provider-specific stop reasons.
     */
    stopReason: v4_1.default.optional(v4_1.default.enum(['endTurn', 'stopSequence', 'maxTokens']).or(v4_1.default.string())),
    role: exports.RoleSchema,
    /**
     * Response content. Single content block (text, image, or audio).
     */
    content: exports.SamplingContentSchema
});
/**
 * The client's response to a sampling/create_message request when tools were provided.
 * This version supports array content for tool use flows.
 */
exports.CreateMessageResultWithToolsSchema = exports.ResultSchema.extend({
    /**
     * The name of the model that generated the message.
     */
    model: v4_1.default.string(),
    /**
     * The reason why sampling stopped, if known.
     *
     * Standard values:
     * - "endTurn": Natural end of the assistant's turn
     * - "stopSequence": A stop sequence was encountered
     * - "maxTokens": Maximum token limit was reached
     * - "toolUse": The model wants to use one or more tools
     *
     * This field is an open string to allow for provider-specific stop reasons.
     */
    stopReason: v4_1.default.optional(v4_1.default.enum(['endTurn', 'stopSequence', 'maxTokens', 'toolUse']).or(v4_1.default.string())),
    role: exports.RoleSchema,
    /**
     * Response content. May be a single block or array. May include ToolUseContent if stopReason is "toolUse".
     */
    content: v4_1.default.union([exports.SamplingMessageContentBlockSchema, v4_1.default.array(exports.SamplingMessageContentBlockSchema)])
});
/* Elicitation */
/**
 * Primitive schema definition for boolean fields.
 */
exports.BooleanSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('boolean'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    default: v4_1.default.boolean().optional()
});
/**
 * Primitive schema definition for string fields.
 */
exports.StringSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('string'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    minLength: v4_1.default.number().optional(),
    maxLength: v4_1.default.number().optional(),
    format: v4_1.default.enum(['email', 'uri', 'date', 'date-time']).optional(),
    default: v4_1.default.string().optional()
});
/**
 * Primitive schema definition for number fields.
 */
exports.NumberSchemaSchema = v4_1.default.object({
    type: v4_1.default.enum(['number', 'integer']),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    minimum: v4_1.default.number().optional(),
    maximum: v4_1.default.number().optional(),
    default: v4_1.default.number().optional()
});
/**
 * Schema for single-selection enumeration without display titles for options.
 */
exports.UntitledSingleSelectEnumSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('string'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    enum: v4_1.default.array(v4_1.default.string()),
    default: v4_1.default.string().optional()
});
/**
 * Schema for single-selection enumeration with display titles for each option.
 */
exports.TitledSingleSelectEnumSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('string'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    oneOf: v4_1.default.array(v4_1.default.object({
        const: v4_1.default.string(),
        title: v4_1.default.string()
    })),
    default: v4_1.default.string().optional()
});
/**
 * Use TitledSingleSelectEnumSchema instead.
 * This interface will be removed in a future version.
 */
exports.LegacyTitledEnumSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('string'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    enum: v4_1.default.array(v4_1.default.string()),
    enumNames: v4_1.default.array(v4_1.default.string()).optional(),
    default: v4_1.default.string().optional()
});
// Combined single selection enumeration
exports.SingleSelectEnumSchemaSchema = v4_1.default.union([exports.UntitledSingleSelectEnumSchemaSchema, exports.TitledSingleSelectEnumSchemaSchema]);
/**
 * Schema for multiple-selection enumeration without display titles for options.
 */
exports.UntitledMultiSelectEnumSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('array'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    minItems: v4_1.default.number().optional(),
    maxItems: v4_1.default.number().optional(),
    items: v4_1.default.object({
        type: v4_1.default.literal('string'),
        enum: v4_1.default.array(v4_1.default.string())
    }),
    default: v4_1.default.array(v4_1.default.string()).optional()
});
/**
 * Schema for multiple-selection enumeration with display titles for each option.
 */
exports.TitledMultiSelectEnumSchemaSchema = v4_1.default.object({
    type: v4_1.default.literal('array'),
    title: v4_1.default.string().optional(),
    description: v4_1.default.string().optional(),
    minItems: v4_1.default.number().optional(),
    maxItems: v4_1.default.number().optional(),
    items: v4_1.default.object({
        anyOf: v4_1.default.array(v4_1.default.object({
            const: v4_1.default.string(),
            title: v4_1.default.string()
        }))
    }),
    default: v4_1.default.array(v4_1.default.string()).optional()
});
/**
 * Combined schema for multiple-selection enumeration
 */
exports.MultiSelectEnumSchemaSchema = v4_1.default.union([exports.UntitledMultiSelectEnumSchemaSchema, exports.TitledMultiSelectEnumSchemaSchema]);
/**
 * Primitive schema definition for enum fields.
 */
exports.EnumSchemaSchema = v4_1.default.union([exports.LegacyTitledEnumSchemaSchema, exports.SingleSelectEnumSchemaSchema, exports.MultiSelectEnumSchemaSchema]);
/**
 * Union of all primitive schema definitions.
 */
exports.PrimitiveSchemaDefinitionSchema = v4_1.default.union([exports.EnumSchemaSchema, exports.BooleanSchemaSchema, exports.StringSchemaSchema, exports.NumberSchemaSchema]);
/**
 * Parameters for an `elicitation/create` request for form-based elicitation.
 */
exports.ElicitRequestFormParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The elicitation mode.
     *
     * Optional for backward compatibility. Clients MUST treat missing mode as "form".
     */
    mode: v4_1.default.literal('form').optional(),
    /**
     * The message to present to the user describing what information is being requested.
     */
    message: v4_1.default.string(),
    /**
     * A restricted subset of JSON Schema.
     * Only top-level properties are allowed, without nesting.
     */
    requestedSchema: v4_1.default.object({
        type: v4_1.default.literal('object'),
        properties: v4_1.default.record(v4_1.default.string(), exports.PrimitiveSchemaDefinitionSchema),
        required: v4_1.default.array(v4_1.default.string()).optional()
    })
});
/**
 * Parameters for an `elicitation/create` request for URL-based elicitation.
 */
exports.ElicitRequestURLParamsSchema = BaseRequestParamsSchema.extend({
    /**
     * The elicitation mode.
     */
    mode: v4_1.default.literal('url'),
    /**
     * The message to present to the user explaining why the interaction is needed.
     */
    message: v4_1.default.string(),
    /**
     * The ID of the elicitation, which must be unique within the context of the server.
     * The client MUST treat this ID as an opaque value.
     */
    elicitationId: v4_1.default.string(),
    /**
     * The URL that the user should navigate to.
     */
    url: v4_1.default.string().url()
});
/**
 * The parameters for a request to elicit additional information from the user via the client.
 */
exports.ElicitRequestParamsSchema = v4_1.default.union([exports.ElicitRequestFormParamsSchema, exports.ElicitRequestURLParamsSchema]);
/**
 * A request from the server to elicit user input via the client.
 * The client should present the message and form fields to the user (form mode)
 * or navigate to a URL (URL mode).
 */
exports.ElicitRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('elicitation/create'),
    params: exports.ElicitRequestParamsSchema
});
/**
 * Parameters for a `notifications/elicitation/complete` notification.
 *
 * @category notifications/elicitation/complete
 */
exports.ElicitationCompleteNotificationParamsSchema = NotificationsParamsSchema.extend({
    /**
     * The ID of the elicitation that completed.
     */
    elicitationId: v4_1.default.string()
});
/**
 * A notification from the server to the client, informing it of a completion of an out-of-band elicitation request.
 *
 * @category notifications/elicitation/complete
 */
exports.ElicitationCompleteNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/elicitation/complete'),
    params: exports.ElicitationCompleteNotificationParamsSchema
});
/**
 * The client's response to an elicitation/create request from the server.
 */
exports.ElicitResultSchema = exports.ResultSchema.extend({
    /**
     * The user action in response to the elicitation.
     * - "accept": User submitted the form/confirmed the action
     * - "decline": User explicitly decline the action
     * - "cancel": User dismissed without making an explicit choice
     */
    action: v4_1.default.enum(['accept', 'decline', 'cancel']),
    /**
     * The submitted form data, only present when action is "accept".
     * Contains values matching the requested schema.
     * Per MCP spec, content is "typically omitted" for decline/cancel actions.
     * We normalize null to undefined for leniency while maintaining type compatibility.
     */
    content: v4_1.default.preprocess(val => (val === null ? undefined : val), v4_1.default.record(v4_1.default.string(), v4_1.default.union([v4_1.default.string(), v4_1.default.number(), v4_1.default.boolean(), v4_1.default.array(v4_1.default.string())])).optional())
});
/* Autocomplete */
/**
 * A reference to a resource or resource template definition.
 */
exports.ResourceTemplateReferenceSchema = v4_1.default.object({
    type: v4_1.default.literal('ref/resource'),
    /**
     * The URI or URI template of the resource.
     */
    uri: v4_1.default.string()
});
/**
 * @deprecated Use ResourceTemplateReferenceSchema instead
 */
exports.ResourceReferenceSchema = exports.ResourceTemplateReferenceSchema;
/**
 * Identifies a prompt.
 */
exports.PromptReferenceSchema = v4_1.default.object({
    type: v4_1.default.literal('ref/prompt'),
    /**
     * The name of the prompt or prompt template
     */
    name: v4_1.default.string()
});
/**
 * Parameters for a `completion/complete` request.
 */
exports.CompleteRequestParamsSchema = BaseRequestParamsSchema.extend({
    ref: v4_1.default.union([exports.PromptReferenceSchema, exports.ResourceTemplateReferenceSchema]),
    /**
     * The argument's information
     */
    argument: v4_1.default.object({
        /**
         * The name of the argument
         */
        name: v4_1.default.string(),
        /**
         * The value of the argument to use for completion matching.
         */
        value: v4_1.default.string()
    }),
    context: v4_1.default
        .object({
        /**
         * Previously-resolved variables in a URI template or prompt.
         */
        arguments: v4_1.default.record(v4_1.default.string(), v4_1.default.string()).optional()
    })
        .optional()
});
/**
 * A request from the client to the server, to ask for completion options.
 */
exports.CompleteRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('completion/complete'),
    params: exports.CompleteRequestParamsSchema
});
function assertCompleteRequestPrompt(request) {
    if (request.params.ref.type !== 'ref/prompt') {
        throw new TypeError(`Expected CompleteRequestPrompt, but got ${request.params.ref.type}`);
    }
    void request;
}
function assertCompleteRequestResourceTemplate(request) {
    if (request.params.ref.type !== 'ref/resource') {
        throw new TypeError(`Expected CompleteRequestResourceTemplate, but got ${request.params.ref.type}`);
    }
    void request;
}
/**
 * The server's response to a completion/complete request
 */
exports.CompleteResultSchema = exports.ResultSchema.extend({
    completion: v4_1.default.looseObject({
        /**
         * An array of completion values. Must not exceed 100 items.
         */
        values: v4_1.default.array(v4_1.default.string()).max(100),
        /**
         * The total number of completion options available. This can exceed the number of values actually sent in the response.
         */
        total: v4_1.default.optional(v4_1.default.number().int()),
        /**
         * Indicates whether there are additional completion options beyond those provided in the current response, even if the exact total is unknown.
         */
        hasMore: v4_1.default.optional(v4_1.default.boolean())
    })
});
/* Roots */
/**
 * Represents a root directory or file that the server can operate on.
 */
exports.RootSchema = v4_1.default.object({
    /**
     * The URI identifying the root. This *must* start with file:// for now.
     */
    uri: v4_1.default.string().startsWith('file://'),
    /**
     * An optional name for the root.
     */
    name: v4_1.default.string().optional(),
    /**
     * See [MCP specification](https://github.com/modelcontextprotocol/modelcontextprotocol/blob/47339c03c143bb4ec01a26e721a1b8fe66634ebe/docs/specification/draft/basic/index.mdx#general-fields)
     * for notes on _meta usage.
     */
    _meta: v4_1.default.record(v4_1.default.string(), v4_1.default.unknown()).optional()
});
/**
 * Sent from the server to request a list of root URIs from the client.
 */
exports.ListRootsRequestSchema = exports.RequestSchema.extend({
    method: v4_1.default.literal('roots/list')
});
/**
 * The client's response to a roots/list request from the server.
 */
exports.ListRootsResultSchema = exports.ResultSchema.extend({
    roots: v4_1.default.array(exports.RootSchema)
});
/**
 * A notification from the client to the server, informing it that the list of roots has changed.
 */
exports.RootsListChangedNotificationSchema = exports.NotificationSchema.extend({
    method: v4_1.default.literal('notifications/roots/list_changed')
});
/* Client messages */
exports.ClientRequestSchema = v4_1.default.union([
    exports.PingRequestSchema,
    exports.InitializeRequestSchema,
    exports.CompleteRequestSchema,
    exports.SetLevelRequestSchema,
    exports.GetPromptRequestSchema,
    exports.ListPromptsRequestSchema,
    exports.ListResourcesRequestSchema,
    exports.ListResourceTemplatesRequestSchema,
    exports.ReadResourceRequestSchema,
    exports.SubscribeRequestSchema,
    exports.UnsubscribeRequestSchema,
    exports.CallToolRequestSchema,
    exports.ListToolsRequestSchema,
    exports.GetTaskRequestSchema,
    exports.GetTaskPayloadRequestSchema,
    exports.ListTasksRequestSchema
]);
exports.ClientNotificationSchema = v4_1.default.union([
    exports.CancelledNotificationSchema,
    exports.ProgressNotificationSchema,
    exports.InitializedNotificationSchema,
    exports.RootsListChangedNotificationSchema,
    exports.TaskStatusNotificationSchema
]);
exports.ClientResultSchema = v4_1.default.union([
    exports.EmptyResultSchema,
    exports.CreateMessageResultSchema,
    exports.CreateMessageResultWithToolsSchema,
    exports.ElicitResultSchema,
    exports.ListRootsResultSchema,
    exports.GetTaskResultSchema,
    exports.ListTasksResultSchema,
    exports.CreateTaskResultSchema
]);
/* Server messages */
exports.ServerRequestSchema = v4_1.default.union([
    exports.PingRequestSchema,
    exports.CreateMessageRequestSchema,
    exports.ElicitRequestSchema,
    exports.ListRootsRequestSchema,
    exports.GetTaskRequestSchema,
    exports.GetTaskPayloadRequestSchema,
    exports.ListTasksRequestSchema
]);
exports.ServerNotificationSchema = v4_1.default.union([
    exports.CancelledNotificationSchema,
    exports.ProgressNotificationSchema,
    exports.LoggingMessageNotificationSchema,
    exports.ResourceUpdatedNotificationSchema,
    exports.ResourceListChangedNotificationSchema,
    exports.ToolListChangedNotificationSchema,
    exports.PromptListChangedNotificationSchema,
    exports.TaskStatusNotificationSchema,
    exports.ElicitationCompleteNotificationSchema
]);
exports.ServerResultSchema = v4_1.default.union([
    exports.EmptyResultSchema,
    exports.InitializeResultSchema,
    exports.CompleteResultSchema,
    exports.GetPromptResultSchema,
    exports.ListPromptsResultSchema,
    exports.ListResourcesResultSchema,
    exports.ListResourceTemplatesResultSchema,
    exports.ReadResourceResultSchema,
    exports.CallToolResultSchema,
    exports.ListToolsResultSchema,
    exports.GetTaskResultSchema,
    exports.ListTasksResultSchema,
    exports.CreateTaskResultSchema
]);
class McpError extends Error {
    constructor(code, message, data) {
        super(`MCP error ${code}: ${message}`);
        this.code = code;
        this.data = data;
        this.name = 'McpError';
    }
    /**
     * Factory method to create the appropriate error type based on the error code and data
     */
    static fromError(code, message, data) {
        // Check for specific error types
        if (code === ErrorCode.UrlElicitationRequired && data) {
            const errorData = data;
            if (errorData.elicitations) {
                return new UrlElicitationRequiredError(errorData.elicitations, message);
            }
        }
        // Default to generic McpError
        return new McpError(code, message, data);
    }
}
exports.McpError = McpError;
/**
 * Specialized error type when a tool requires a URL mode elicitation.
 * This makes it nicer for the client to handle since there is specific data to work with instead of just a code to check against.
 */
class UrlElicitationRequiredError extends McpError {
    constructor(elicitations, message = `URL elicitation${elicitations.length > 1 ? 's' : ''} required`) {
        super(ErrorCode.UrlElicitationRequired, message, {
            elicitations: elicitations
        });
    }
    get elicitations() {
        var _a, _b;
        return (_b = (_a = this.data) === null || _a === void 0 ? void 0 : _a.elicitations) !== null && _b !== void 0 ? _b : [];
    }
}
exports.UrlElicitationRequiredError = UrlElicitationRequiredError;
